# Pair Unión y Limpieza de Datos

Para realizar estos ejercicios deberéis usar el conjunto de datos de world-data-2023-part1.csv y el de world-data-2023-part2.csv.

## 1. Ejercicios Unión de Datos

Tienes a tu disposición dos conjuntos de datos, "world-data-2023-part1.csv" y "world-data-2023-part2.csv", que contienen información de una serie de indicadores y datos de distintos países. Tu tarea es explorar estos conjuntos de datos y determinar qué tienen en común en términos de columnas y datos.

Luego, debes crear un nuevo DataFrame que combine la información de ambos conjuntos de datos en un solo conjunto de datos. Para hacerlo, debes seleccionar el método de unión de Pandas que consideres más apropiado para esta situación y justificar por qué crees que ese método es el mejor en tu informe.

Asegúrate de realizar los siguientes pasos:

Explora y carga ambos conjuntos de datos en pandas DataFrames.

Identifica las columnas comunes entre los dos conjuntos de datos.

Utiliza el método de unión de Pandas que consideres más adecuado para combinar los datos de ambos años en un solo DataFrame.

Explica por qué elegiste ese método de unión y cómo se llevaron a cabo los pasos anteriores.


In [159]:
# importamos las librerías que necesitamos

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np


# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

In [160]:
df1 = pd.read_csv("files/world-data-2023_part1.csv", index_col= 0)
df2 = pd.read_csv("files/world-data-2023_part2.csv", index_col= 0)

In [161]:
df1.head()

Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0  Afghanistan               60           AF                58.10%   
1      Albania              105           AL                43.10%   
2      Algeria               18           DZ                17.40%   
3      Andorra              164           AD                40.00%   
4       Angola               26           AO                47.50%   

  Land Area(Km2) Armed Forces size  Birth Rate  Calling Code  \
0        652,230           323,000       32.49          93.0   
1         28,748             9,000       11.78         355.0   
2      2,381,741           317,000       24.28         213.0   
3            468               NaN        7.20         376.0   
4      1,246,700           117,000       40.73         244.0   

  Capital/Major City Co2-Emissions     CPI CPI Change (%) Currency-Code  \
0              Kabul         8,672   149.9          2.30%           AFN   
1             Tirana         4,536  119.05          1.40%           ALL   
2            Algiers       150,006  151.36          2.00%           DZD   
3   Andorra la Vella           469     NaN            NaN           EUR   
4             Luanda        34,693  261.73         17.10%           AOA   

   Fertility Rate Forested Area (%) Gasoline Price  
0            4.47             2.10%         $0.70   
1            1.62            28.10%         $1.36   
2            3.02             0.80%         $0.28   
3            1.27            34.00%         $1.51   
4            5.52            46.30%         $0.97

In [162]:
df2.head()

GDP Gross primary education enrollment (%)  \
0   $19,101,353,833                                 104.00%   
1   $15,278,077,447                                 107.00%   
2  $169,988,236,398                                 109.90%   
3    $3,154,057,987                                 106.40%   
4   $94,635,415,870                                 113.50%   

  Gross tertiary education enrollment (%)  Infant mortality      Largest city  \
0                                   9.70%              47.9             Kabul   
1                                  55.00%               7.8            Tirana   
2                                  51.40%              20.1           Algiers   
3                                     NaN               2.7  Andorra la Vella   
4                                   9.30%              51.6            Luanda   

   Life expectancy  Maternal mortality ratio Minimum wage Official language  \
0             64.5                     638.0       $0.43             Pashto   
1             78.5                      15.0       $1.12           Albanian   
2             76.7                     112.0       $0.95             Arabic   
3              NaN                       NaN       $6.63            Catalan   
4             60.8                     241.0       $0.71         Portuguese   

  Out of pocket health expenditure  Physicians per thousand  Population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
0                                    48.90%           9.30%         71.40%   
1                                    55.70%          18.60%         36.60%   
2                                    41.20%          37.20%         66.10%   
3                                       NaN             NaN            NaN   
4                                    77.50%           9.20%         49.10%   

  Unemployment rate Urban_population      country  \
0            11.12%        9,797,273  Afghanistan   
1            12.33%        1,747,593      Albania   
2            11.70%       31,510,100      Algeria   
3               NaN           67,873      Andorra   
4             6.89%       21,061,025       Angola   

                    coordinates  
0    ('33.93911 ', '67.709953')  
1   ('41.153332 ', '20.168331')  
2    ('28.033886 ', '1.659626')  
3    ('42.506285 ', '1.521801')  
4  ('-11.202692 ', '17.873887')

Parece que ambas tablas tienen en común la columna "Country". Vamos a investigar más sobre esta columna en ambos df.

In [163]:
df1.shape #compruebo si tienen el mimo número de países (países = num filas, en este caso parece q 195)

(195, 16)

In [164]:
df1["Country"].nunique() #y no se repiten

195

In [165]:
df2.shape #tienen 195 filas (=países)

(195, 19)

In [166]:
df2["country"].nunique() #y no se repiten

195

Para juntar los datos:

- ambos df coinciden en una columna = hay que hacer un `merge`. (nota: si coincidieran en el índice, habría que hacer `join`).

- tipo de merge: parece a simple vista que daría igual hacer un `inner` que un `left`, porque entendemos que son los mismos países en ambos dataframe. Si no me fiara de quien me diera los datos igual sí haría un bucle for para comparar si ambos datos son iguales, como por ejemplo, el siguiente:

(aunque entendemos que esto no es necesario)

In [167]:
coincidencias = []
disparidades = {}

for x,y in zip(df1['Country'], df2['country']): #parece que ambas columnas están ordenadas por orden alfabético
    if x == y:
        coincidencias.append(x)
    else:
        disparidades['x'] = y


#print('coincidencias:', coincidencias)
print(len(coincidencias))

print('disparidades:', disparidades)
print(len(disparidades))

195
disparidades: {}
0


In [168]:
# hacemos el merge. Por defecto coge el "inner", así que no le diremos nada

df_merge = df1.merge(df2, left_on = "Country", right_on = "country")

In [169]:
df_merge.head()

Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0  Afghanistan               60           AF                58.10%   
1      Albania              105           AL                43.10%   
2      Algeria               18           DZ                17.40%   
3      Andorra              164           AD                40.00%   
4       Angola               26           AO                47.50%   

  Land Area(Km2) Armed Forces size  Birth Rate  Calling Code  \
0        652,230           323,000       32.49          93.0   
1         28,748             9,000       11.78         355.0   
2      2,381,741           317,000       24.28         213.0   
3            468               NaN        7.20         376.0   
4      1,246,700           117,000       40.73         244.0   

  Capital/Major City Co2-Emissions     CPI CPI Change (%) Currency-Code  \
0              Kabul         8,672   149.9          2.30%           AFN   
1             Tirana         4,536  119.05          1.40%           ALL   
2            Algiers       150,006  151.36          2.00%           DZD   
3   Andorra la Vella           469     NaN            NaN           EUR   
4             Luanda        34,693  261.73         17.10%           AOA   

   Fertility Rate Forested Area (%) Gasoline Price                GDP  \
0            4.47             2.10%         $0.70    $19,101,353,833    
1            1.62            28.10%         $1.36    $15,278,077,447    
2            3.02             0.80%         $0.28   $169,988,236,398    
3            1.27            34.00%         $1.51     $3,154,057,987    
4            5.52            46.30%         $0.97    $94,635,415,870    

  Gross primary education enrollment (%)  \
0                                104.00%   
1                                107.00%   
2                                109.90%   
3                                106.40%   
4                                113.50%   

  Gross tertiary education enrollment (%)  Infant mortality      Largest city  \
0                                   9.70%              47.9             Kabul   
1                                  55.00%               7.8            Tirana   
2                                  51.40%              20.1           Algiers   
3                                     NaN               2.7  Andorra la Vella   
4                                   9.30%              51.6            Luanda   

   Life expectancy  Maternal mortality ratio Minimum wage Official language  \
0             64.5                     638.0       $0.43             Pashto   
1             78.5                      15.0       $1.12           Albanian   
2             76.7                     112.0       $0.95             Arabic   
3              NaN                       NaN       $6.63            Catalan   
4             60.8                     241.0       $0.71         Portuguese   

  Out of pocket health expenditure  Physicians per thousand  Population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
0                                    48.90%           9.30%         71.40%   
1                                    55.70%          18.60%         36.60%   
2                                    41.20%          37.20%         66.10%   
3                                       NaN             NaN            NaN   
4                                    77.50%           9.20%         49.10%   

  Unemployment rate Urban_population      country  \
0            11.12%        9,797,273  Afghanistan   
1            12.33%        1,747,593      Albania   
2    

## 2. Ejercicios de Limpieza

1. Después de la unión de datos, tenemos dos columnas de "country". Elimina una de ellas.

2. Los nombres de las columnas no son homogeneos. Cambia los nombres de las columnas de tal forma que:

- No tengan espacios.

- Estén en minúscula.

- No tengan paréntesis, es decir, quitar "(%)", "(Km2)".

- Algunas columnas tiene "\n". Eliminalos de los nombres de las columnas.

- Algunas columnas tienen ":". Eliminalos de los nombres de las columnas.

3. La columnas coordinates tiene la latitud y la longitud en una sola columna. Crea dos columnas nuevas, una con la longitud y otra con la latitud. Una vez hecho, elimina la columna de coordinates.

4. Las columnas unemployment_rate, total_tax_rate, tax_revenue, population_labor_force_participation, out_of_pocket_health_expenditure, gross_tertiary_education_enrollment, gross_primary_education_enrollment, forested_area, cpi_change, agricultural_land tienen "%". Elimina los "%" de los valores de las columnas.

5. Haz lo mismo para las columnas de gasoline_price, gdp, minimum_wage, pero eliminando "$".

6. Guarda el DataFrame para usarlo posteriormente.


In [170]:
#Después de la unión de datos, tenemos dos columnas de "country". Elimina una de ellas.
#voy a eliminar la que aparece en minúsculas, que es la que está al final del df:

df_merge.drop(columns = ['country'], inplace = True)

In [171]:
#compruebo:
df_merge.head()

Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0  Afghanistan               60           AF                58.10%   
1      Albania              105           AL                43.10%   
2      Algeria               18           DZ                17.40%   
3      Andorra              164           AD                40.00%   
4       Angola               26           AO                47.50%   

  Land Area(Km2) Armed Forces size  Birth Rate  Calling Code  \
0        652,230           323,000       32.49          93.0   
1         28,748             9,000       11.78         355.0   
2      2,381,741           317,000       24.28         213.0   
3            468               NaN        7.20         376.0   
4      1,246,700           117,000       40.73         244.0   

  Capital/Major City Co2-Emissions     CPI CPI Change (%) Currency-Code  \
0              Kabul         8,672   149.9          2.30%           AFN   
1             Tirana         4,536  119.05          1.40%           ALL   
2            Algiers       150,006  151.36          2.00%           DZD   
3   Andorra la Vella           469     NaN            NaN           EUR   
4             Luanda        34,693  261.73         17.10%           AOA   

   Fertility Rate Forested Area (%) Gasoline Price                GDP  \
0            4.47             2.10%         $0.70    $19,101,353,833    
1            1.62            28.10%         $1.36    $15,278,077,447    
2            3.02             0.80%         $0.28   $169,988,236,398    
3            1.27            34.00%         $1.51     $3,154,057,987    
4            5.52            46.30%         $0.97    $94,635,415,870    

  Gross primary education enrollment (%)  \
0                                104.00%   
1                                107.00%   
2                                109.90%   
3                                106.40%   
4                                113.50%   

  Gross tertiary education enrollment (%)  Infant mortality      Largest city  \
0                                   9.70%              47.9             Kabul   
1                                  55.00%               7.8            Tirana   
2                                  51.40%              20.1           Algiers   
3                                     NaN               2.7  Andorra la Vella   
4                                   9.30%              51.6            Luanda   

   Life expectancy  Maternal mortality ratio Minimum wage Official language  \
0             64.5                     638.0       $0.43             Pashto   
1             78.5                      15.0       $1.12           Albanian   
2             76.7                     112.0       $0.95             Arabic   
3              NaN                       NaN       $6.63            Catalan   
4             60.8                     241.0       $0.71         Portuguese   

  Out of pocket health expenditure  Physicians per thousand  Population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  Population: Labor force participation (%) Tax revenue (%) Total tax rate  \
0                                    48.90%           9.30%         71.40%   
1                                    55.70%          18.60%         36.60%   
2                                    41.20%          37.20%         66.10%   
3                                       NaN             NaN            NaN   
4                                    77.50%           9.20%         49.10%   

  Unemployment rate Urban_population                   coordinates  
0            11.12%        9,797,273    ('33.93911 ', '67.709953')  
1            12.33%       

Limpieza de nombres en las columnas:


- str.strip() = quita los espacios al principio y al final del string (strig = nombre columna)

- str.lower() = pone el string en minus

- str.split() = separa el string en un función del separador que le des. En este caso, cuando haya paréntesis, le decimos que se quede con todo lo que hay antes del paréntesis. Nos devolverá una lista dividida por el separador "(" Ej:Agricultural Land( %) =  [Agricultural Land, ( %)]. Nos quedaremos sólo con Agricultural Land con el `lista[0]`

- str.replace() = para los "\n" y ":". Nos reemplaza ese string por un espacio. 

In [172]:
#primero hago una lista de columnas:
lista_columnas = [col for col in df_merge.columns]

lista_columnas[:3]#que nos muestre los 3 primeros elementos, para ver si está ok.

['Country', 'Density\n(P/Km2)', 'Abbreviation']

In [173]:
#vamos a hacer una nueva lista de columnas limpia, para luego poder hacer el rename.
lista_columnas_limpia = []

for col in lista_columnas:
    col1 = col.strip().lower().split("(")
    col2 = str(col1[0]) #he tenido q convertir en lista pq las listas no tienen la propiedad replace que quiero aplicarle en el siguiente paso. sólo los strings
    
     #reemplazamos: 
    col3 = col2.replace('\n', '').replace(':', ' ').strip().replace(' ', '_')
    
    lista_columnas_limpia.append(col3)

print(lista_columnas_limpia)

['country', 'density', 'abbreviation', 'agricultural_land', 'land_area', 'armed_forces_size', 'birth_rate', 'calling_code', 'capital/major_city', 'co2-emissions', 'cpi', 'cpi_change', 'currency-code', 'fertility_rate', 'forested_area', 'gasoline_price', 'gdp', 'gross_primary_education_enrollment', 'gross_tertiary_education_enrollment', 'infant_mortality', 'largest_city', 'life_expectancy', 'maternal_mortality_ratio', 'minimum_wage', 'official_language', 'out_of_pocket_health_expenditure', 'physicians_per_thousand', 'population', 'population__labor_force_participation', 'tax_revenue', 'total_tax_rate', 'unemployment_rate', 'urban_population', 'coordinates']


In [174]:
#el rename sólo puede hacerse con un diccionario==> {antiguo valor: nuevo valor}
#por eso vamos a crear ese diccionario con la función zip():

dicc_col = {}

for x,y in zip(lista_columnas, lista_columnas_limpia):
    dicc_col[x] = y

In [175]:
#renombramos las columnas:

df_merge.rename(columns=dicc_col, inplace = True)

In [176]:
df_merge.head()

country density abbreviation agricultural_land  land_area  \
0  Afghanistan      60           AF            58.10%    652,230   
1      Albania     105           AL            43.10%     28,748   
2      Algeria      18           DZ            17.40%  2,381,741   
3      Andorra     164           AD            40.00%        468   
4       Angola      26           AO            47.50%  1,246,700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2-emissions     cpi cpi_change currency-code  fertility_rate  \
0         8,672   149.9      2.30%           AFN            4.47   
1         4,536  119.05      1.40%           ALL            1.62   
2       150,006  151.36      2.00%           DZD            3.02   
3           469     NaN        NaN           EUR            1.27   
4        34,693  261.73     17.10%           AOA            5.52   

  forested_area gasoline_price                gdp  \
0         2.10%         $0.70    $19,101,353,833    
1        28.10%         $1.36    $15,278,077,447    
2         0.80%         $0.28   $169,988,236,398    
3        34.00%         $1.51     $3,154,057,987    
4        46.30%         $0.97    $94,635,415,870    

  gross_primary_education_enrollment gross_tertiary_education_enrollment  \
0                            104.00%                               9.70%   
1                            107.00%                              55.00%   
2                            109.90%                              51.40%   
3                            106.40%                                 NaN   
4                            113.50%                               9.30%   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio minimum_wage official_language  \
0                     638.0       $0.43             Pashto   
1                      15.0       $1.12           Albanian   
2                     112.0       $0.95             Arabic   
3                       NaN       $6.63            Catalan   
4                     241.0       $0.71         Portuguese   

  out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  population__labor_force_participation tax_revenue total_tax_rate  \
0                                48.90%       9.30%         71.40%   
1                                55.70%      18.60%         36.60%   
2                                41.20%      37.20%         66.10%   
3                                   NaN         NaN            NaN   
4                                77.50%       9.20%         49.10%   

  unemployment_rate urban_population                   coordinates  
0            11.12%        9,797,273    ('33.93911 ', '67.709953')  
1            12.33%        1,747,593   ('41.153332 ', '20.168331')  
2            11.70%       31,510,100    ('28.033886 ', '1.659626')  
3               NaN           67,873    ('42.506285 ', '1.521801')  
4             6.89%       21,061,025  ('-11.202692 ', '17.873887'

La columna `coordinates` tiene la latitud y la longitud en una sola columna. Crea dos columnas nuevas, una con la longitud y otra con la latitud. Una vez hecho, elimina la columna de coordinates.

Para ello, utilizamos: 
```python
df[nombre_columna].str.split(',', expand = True).get(0,1)
```

- `.str`= permite aplicar a los valores de una columna el método de strings que le digas, en este caso, utilizamos el `.split()`

- `.split()`: indicasvalor a reemplazar y el valor por el cual quieres reemplazar. El split te transforma el valor en una lista. El `spand = True` nos crea una columna para cada elemento de la lista generado por el split: ["('33.93911 '", " '67.709953')"]. Nos creará dos columnas.




In [177]:
#el código anterior, habrá que igualarlo a dos nuevas columnas, para que aparezcan en el df:

df_merge[['latitud', 'longitud']] = df_merge['coordinates'].str.split(',', expand = True)

In [178]:
df_merge.head() #ahora tendremos que limpiar los datos de estas columnas nuevas.

country density abbreviation agricultural_land  land_area  \
0  Afghanistan      60           AF            58.10%    652,230   
1      Albania     105           AL            43.10%     28,748   
2      Algeria      18           DZ            17.40%  2,381,741   
3      Andorra     164           AD            40.00%        468   
4       Angola      26           AO            47.50%  1,246,700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2-emissions     cpi cpi_change currency-code  fertility_rate  \
0         8,672   149.9      2.30%           AFN            4.47   
1         4,536  119.05      1.40%           ALL            1.62   
2       150,006  151.36      2.00%           DZD            3.02   
3           469     NaN        NaN           EUR            1.27   
4        34,693  261.73     17.10%           AOA            5.52   

  forested_area gasoline_price                gdp  \
0         2.10%         $0.70    $19,101,353,833    
1        28.10%         $1.36    $15,278,077,447    
2         0.80%         $0.28   $169,988,236,398    
3        34.00%         $1.51     $3,154,057,987    
4        46.30%         $0.97    $94,635,415,870    

  gross_primary_education_enrollment gross_tertiary_education_enrollment  \
0                            104.00%                               9.70%   
1                            107.00%                              55.00%   
2                            109.90%                              51.40%   
3                            106.40%                                 NaN   
4                            113.50%                               9.30%   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio minimum_wage official_language  \
0                     638.0       $0.43             Pashto   
1                      15.0       $1.12           Albanian   
2                     112.0       $0.95             Arabic   
3                       NaN       $6.63            Catalan   
4                     241.0       $0.71         Portuguese   

  out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  population__labor_force_participation tax_revenue total_tax_rate  \
0                                48.90%       9.30%         71.40%   
1                                55.70%      18.60%         36.60%   
2                                41.20%      37.20%         66.10%   
3                                   NaN         NaN            NaN   
4                                77.50%       9.20%         49.10%   

  unemployment_rate urban_population                   coordinates  \
0            11.12%        9,797,273    ('33.93911 ', '67.709953')   
1            12.33%        1,747,593   ('41.153332 ', '20.168331')   
2            11.70%       31,510,100    ('28.033886 ', '1.659626')   
3               NaN           67,873    ('42.506285 ', '1.521801')   
4             6.89%       21,061,025  ('-11.202692 ', '17.87

In [179]:
df_merge['latitud'] = df_merge['latitud'].str.replace('(', '').str.replace(' ', '').str.replace("'", "")
df_merge['latitud'] = df_merge['latitud'].str.strip() 
#parece que este .strip no lo coge, esto es porque el dato no es de tipo string, sino de tipo object.
#voy a intentar cambiarlo a string en el siguiente pasito
df_merge.head()

country density abbreviation agricultural_land  land_area  \
0  Afghanistan      60           AF            58.10%    652,230   
1      Albania     105           AL            43.10%     28,748   
2      Algeria      18           DZ            17.40%  2,381,741   
3      Andorra     164           AD            40.00%        468   
4       Angola      26           AO            47.50%  1,246,700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2-emissions     cpi cpi_change currency-code  fertility_rate  \
0         8,672   149.9      2.30%           AFN            4.47   
1         4,536  119.05      1.40%           ALL            1.62   
2       150,006  151.36      2.00%           DZD            3.02   
3           469     NaN        NaN           EUR            1.27   
4        34,693  261.73     17.10%           AOA            5.52   

  forested_area gasoline_price                gdp  \
0         2.10%         $0.70    $19,101,353,833    
1        28.10%         $1.36    $15,278,077,447    
2         0.80%         $0.28   $169,988,236,398    
3        34.00%         $1.51     $3,154,057,987    
4        46.30%         $0.97    $94,635,415,870    

  gross_primary_education_enrollment gross_tertiary_education_enrollment  \
0                            104.00%                               9.70%   
1                            107.00%                              55.00%   
2                            109.90%                              51.40%   
3                            106.40%                                 NaN   
4                            113.50%                               9.30%   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio minimum_wage official_language  \
0                     638.0       $0.43             Pashto   
1                      15.0       $1.12           Albanian   
2                     112.0       $0.95             Arabic   
3                       NaN       $6.63            Catalan   
4                     241.0       $0.71         Portuguese   

  out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  population__labor_force_participation tax_revenue total_tax_rate  \
0                                48.90%       9.30%         71.40%   
1                                55.70%      18.60%         36.60%   
2                                41.20%      37.20%         66.10%   
3                                   NaN         NaN            NaN   
4                                77.50%       9.20%         49.10%   

  unemployment_rate urban_population                   coordinates  \
0            11.12%        9,797,273    ('33.93911 ', '67.709953')   
1            12.33%        1,747,593   ('41.153332 ', '20.168331')   
2            11.70%       31,510,100    ('28.033886 ', '1.659626')   
3               NaN           67,873    ('42.506285 ', '1.521801')   
4             6.89%       21,061,025  ('-11.202692 ', '17.87

In [180]:
#ahora lo hago con la otra columna:
df_merge['longitud'] = df_merge['longitud'].str.replace(')', '').str.replace(' ', '').str.replace("'", "")
df_merge.head()

country density abbreviation agricultural_land  land_area  \
0  Afghanistan      60           AF            58.10%    652,230   
1      Albania     105           AL            43.10%     28,748   
2      Algeria      18           DZ            17.40%  2,381,741   
3      Andorra     164           AD            40.00%        468   
4       Angola      26           AO            47.50%  1,246,700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2-emissions     cpi cpi_change currency-code  fertility_rate  \
0         8,672   149.9      2.30%           AFN            4.47   
1         4,536  119.05      1.40%           ALL            1.62   
2       150,006  151.36      2.00%           DZD            3.02   
3           469     NaN        NaN           EUR            1.27   
4        34,693  261.73     17.10%           AOA            5.52   

  forested_area gasoline_price                gdp  \
0         2.10%         $0.70    $19,101,353,833    
1        28.10%         $1.36    $15,278,077,447    
2         0.80%         $0.28   $169,988,236,398    
3        34.00%         $1.51     $3,154,057,987    
4        46.30%         $0.97    $94,635,415,870    

  gross_primary_education_enrollment gross_tertiary_education_enrollment  \
0                            104.00%                               9.70%   
1                            107.00%                              55.00%   
2                            109.90%                              51.40%   
3                            106.40%                                 NaN   
4                            113.50%                               9.30%   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio minimum_wage official_language  \
0                     638.0       $0.43             Pashto   
1                      15.0       $1.12           Albanian   
2                     112.0       $0.95             Arabic   
3                       NaN       $6.63            Catalan   
4                     241.0       $0.71         Portuguese   

  out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  population__labor_force_participation tax_revenue total_tax_rate  \
0                                48.90%       9.30%         71.40%   
1                                55.70%      18.60%         36.60%   
2                                41.20%      37.20%         66.10%   
3                                   NaN         NaN            NaN   
4                                77.50%       9.20%         49.10%   

  unemployment_rate urban_population                   coordinates  \
0            11.12%        9,797,273    ('33.93911 ', '67.709953')   
1            12.33%        1,747,593   ('41.153332 ', '20.168331')   
2            11.70%       31,510,100    ('28.033886 ', '1.659626')   
3               NaN           67,873    ('42.506285 ', '1.521801')   
4             6.89%       21,061,025  ('-11.202692 ', '17.87

In [181]:
#ahora, elimino la columna coordinates:
df_merge.drop(columns = ['coordinates'],inplace = True)

In [182]:
df_merge.head()

country density abbreviation agricultural_land  land_area  \
0  Afghanistan      60           AF            58.10%    652,230   
1      Albania     105           AL            43.10%     28,748   
2      Algeria      18           DZ            17.40%  2,381,741   
3      Andorra     164           AD            40.00%        468   
4       Angola      26           AO            47.50%  1,246,700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2-emissions     cpi cpi_change currency-code  fertility_rate  \
0         8,672   149.9      2.30%           AFN            4.47   
1         4,536  119.05      1.40%           ALL            1.62   
2       150,006  151.36      2.00%           DZD            3.02   
3           469     NaN        NaN           EUR            1.27   
4        34,693  261.73     17.10%           AOA            5.52   

  forested_area gasoline_price                gdp  \
0         2.10%         $0.70    $19,101,353,833    
1        28.10%         $1.36    $15,278,077,447    
2         0.80%         $0.28   $169,988,236,398    
3        34.00%         $1.51     $3,154,057,987    
4        46.30%         $0.97    $94,635,415,870    

  gross_primary_education_enrollment gross_tertiary_education_enrollment  \
0                            104.00%                               9.70%   
1                            107.00%                              55.00%   
2                            109.90%                              51.40%   
3                            106.40%                                 NaN   
4                            113.50%                               9.30%   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio minimum_wage official_language  \
0                     638.0       $0.43             Pashto   
1                      15.0       $1.12           Albanian   
2                     112.0       $0.95             Arabic   
3                       NaN       $6.63            Catalan   
4                     241.0       $0.71         Portuguese   

  out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                           78.40%                     0.28  38,041,754   
1                           56.90%                     1.20   2,854,191   
2                           28.10%                     1.72  43,053,054   
3                           36.40%                     3.33      77,142   
4                           33.40%                     0.21  31,825,295   

  population__labor_force_participation tax_revenue total_tax_rate  \
0                                48.90%       9.30%         71.40%   
1                                55.70%      18.60%         36.60%   
2                                41.20%      37.20%         66.10%   
3                                   NaN         NaN            NaN   
4                                77.50%       9.20%         49.10%   

  unemployment_rate urban_population     latitud   longitud  
0            11.12%        9,797,273    33.93911  67.709953  
1            12.33%        1,747,593   41.153332  20.168331  
2            11.70%       31,510,100   28.033886   1.659626  
3               NaN           67,873   42.506285   1.521801  
4             6.89%       21,061,025  -11.202692  17.873887

In [183]:
df_columns = df_merge.columns

Hago últimos cambios qeu veo que quedan:
- "-" reemplazar por "_"
- veo que hay un "__" lo voy a reemplazar por un "_"

In [184]:
#vamos a hacer una nueva lista de columnas limpia, para luego poder hacer el rename.
lista_columnas_limpia = []

for col in df_columns:
    col1 = col.strip().replace("-", "_").replace("__", "_")
    lista_columnas_limpia.append(col1)

print(lista_columnas_limpia)

['country', 'density', 'abbreviation', 'agricultural_land', 'land_area', 'armed_forces_size', 'birth_rate', 'calling_code', 'capital/major_city', 'co2_emissions', 'cpi', 'cpi_change', 'currency_code', 'fertility_rate', 'forested_area', 'gasoline_price', 'gdp', 'gross_primary_education_enrollment', 'gross_tertiary_education_enrollment', 'infant_mortality', 'largest_city', 'life_expectancy', 'maternal_mortality_ratio', 'minimum_wage', 'official_language', 'out_of_pocket_health_expenditure', 'physicians_per_thousand', 'population', 'population_labor_force_participation', 'tax_revenue', 'total_tax_rate', 'unemployment_rate', 'urban_population', 'latitud', 'longitud']


In [185]:
dicccionario_nuevo = {}
for x,y in zip(df_columns, lista_columnas_limpia):
    dicccionario_nuevo[x] = y

In [186]:
df_merge.rename(columns = dicccionario_nuevo, inplace=True)

Las columnas unemployment_rate, total_tax_rate, tax_revenue, population_labor_force_participation, out_of_pocket_health_expenditure, gross_tertiary_education_enrollment, gross_primary_education_enrollment, forested_area, cpi_change, agricultural_land tienen "%". Elimina los "%" de los valores de las columnas.

In [187]:
cambiar_porcentajes = ["unemployment_rate", "total_tax_rate", "tax_revenue", "population_labor_force_participation", 
                    "out_of_pocket_health_expenditure", "gross_tertiary_education_enrollment", "gross_primary_education_enrollment", 
                    "forested_area", "cpi_change", "agricultural_land"]

In [188]:
for column in cambiar_porcentajes:
    df_merge[column] = df_merge[column].str.replace('%', '') #recordemos que hay siempre que almacenar en variable para que se guarde

In [189]:
for column in cambiar_porcentajes: #para comprobar si se ha hecho bien
    display(df_merge.loc[:2, [column]])

unemployment_rate
0             11.12
1             12.33
2             11.70

total_tax_rate
0          71.40
1          36.60
2          66.10

tax_revenue
0        9.30
1       18.60
2       37.20

population_labor_force_participation
0                                48.90
1                                55.70
2                                41.20

out_of_pocket_health_expenditure
0                            78.40
1                            56.90
2                            28.10

gross_tertiary_education_enrollment
0                                9.70
1                               55.00
2                               51.40

gross_primary_education_enrollment
0                             104.00
1                             107.00
2                             109.90

forested_area
0          2.10
1         28.10
2          0.80

cpi_change
0       2.30
1       1.40
2       2.00

agricultural_land
0             58.10
1             43.10
2             17.40

Haz lo mismo para las columnas de gasoline_price, gdp, minimum_wage, pero eliminando "$".



In [190]:
cambiar_dolar = ["gasoline_price", "gdp", "minimum_wage"]

In [191]:
for column in cambiar_dolar:
    df_merge[column] = df_merge[column].str.replace('$', '') #recordemos que hay siempre que almacenar en variable para que se guarde

In [192]:
for column in cambiar_dolar: #para comprobar si se ha hecho bien
    display(df_merge.loc[:2, [column]])

gasoline_price
0          0.70 
1          1.36 
2          0.28

gdp
0   19,101,353,833 
1   15,278,077,447 
2  169,988,236,398

minimum_wage
0        0.43 
1        1.12 
2        0.95

Guarda el DataFrame para usarlo posteriormente.

In [193]:
df_merge.to_csv('files/word-data-merged.csv', index=False)

## 3. Ejercicios de Filtrado

Encuentra todos los países cuya mortalidad infantil esté entre 40 y 50 personas por kilómetro cuadrado.

In [194]:
#nos volvemos a traer el csv:
df_merge = pd.read_csv('files/word-data-merged.csv')
df_merge.head()

country density abbreviation  agricultural_land  land_area  \
0  Afghanistan      60           AF               58.1    652,230   
1      Albania     105           AL               43.1     28,748   
2      Algeria      18           DZ               17.4  2,381,741   
3      Andorra     164           AD               40.0        468   
4       Angola      26           AO               47.5  1,246,700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2_emissions     cpi  cpi_change currency_code  fertility_rate  \
0         8,672   149.9         2.3           AFN            4.47   
1         4,536  119.05         1.4           ALL            1.62   
2       150,006  151.36         2.0           DZD            3.02   
3           469     NaN         NaN           EUR            1.27   
4        34,693  261.73        17.1           AOA            5.52   

   forested_area  gasoline_price               gdp  \
0            2.1            0.70   19,101,353,833    
1           28.1            1.36   15,278,077,447    
2            0.8            0.28  169,988,236,398    
3           34.0            1.51    3,154,057,987    
4           46.3            0.97   94,635,415,870    

   gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
0                               104.0                                  9.7   
1                               107.0                                 55.0   
2                               109.9                                 51.4   
3                               106.4                                  NaN   
4                               113.5                                  9.3   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio  minimum_wage official_language  \
0                     638.0          0.43            Pashto   
1                      15.0          1.12          Albanian   
2                     112.0          0.95            Arabic   
3                       NaN          6.63           Catalan   
4                     241.0          0.71        Portuguese   

   out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                              78.4                     0.28  38,041,754   
1                              56.9                     1.20   2,854,191   
2                              28.1                     1.72  43,053,054   
3                              36.4                     3.33      77,142   
4                              33.4                     0.21  31,825,295   

   population_labor_force_participation  tax_revenue  total_tax_rate  \
0                                  48.9          9.3            71.4   
1                                  55.7         18.6            36.6   
2                                  41.2         37.2            66.1   
3                                   NaN          NaN             NaN   
4                                  77.5          9.2            49.1   

   unemployment_rate urban_population    latitud   longitud  
0              11.12        9,797,273  33.939110  67.709953  
1              12.33        1,747,593  41.153332  20.168331  
2              11.70       31,510,100  28.033886   1.659626  
3                NaN           67,873  42.506285   1.521801  
4               6.89       21,061,025 -11.2026

`Notas Sansi`: 
- Land Area(Km2)=> Área total de tierra del país en kilómetros cuadrados.
- infant_mortality = Número de muertes por cada 1,000 nacidos vivos antes de cumplir un año de edad.
- Density\n(P/Km2): Densidad de población medida en personas por kilómetro cuadrado.


In [195]:
#vamos a ver qué tipo de datos tenemos, necesitamos float/integer:
df_merge['infant_mortality'].info()
print('-'*40)
df_merge["land_area"].info() #vemos que este segundo es un object, tenemos que pasarlo a string
print('-'*40)
df_merge["density"].info()#vemos que este segundo es un object, tenemos que pasarlo a string

<class 'pandas.core.series.Series'>
RangeIndex: 195 entries, 0 to 194
Series name: infant_mortality
Non-Null Count  Dtype  
--------------  -----  
189 non-null    float64
dtypes: float64(1)
memory usage: 1.7 KB
----------------------------------------
<class 'pandas.core.series.Series'>
RangeIndex: 195 entries, 0 to 194
Series name: land_area
Non-Null Count  Dtype 
--------------  ----- 
194 non-null    object
dtypes: object(1)
memory usage: 1.7+ KB
----------------------------------------
<class 'pandas.core.series.Series'>
RangeIndex: 195 entries, 0 to 194
Series name: density
Non-Null Count  Dtype 
--------------  ----- 
195 non-null    object
dtypes: object(1)
memory usage: 1.7+ KB


In [196]:
df_merge['land_area'] = df_merge['land_area'].str.replace(',', '')

In [197]:
df_merge['density'] = df_merge['density'].str.replace(',', '')

In [198]:
df_merge["infant_mortality"].describe()

count    189.000000
mean      21.332804
std       19.548058
min        1.400000
25%        6.000000
50%       14.000000
75%       32.700000
max       84.500000
Name: infant_mortality, dtype: float64

In [199]:
df_merge.head()

country density abbreviation  agricultural_land land_area  \
0  Afghanistan      60           AF               58.1    652230   
1      Albania     105           AL               43.1     28748   
2      Algeria      18           DZ               17.4   2381741   
3      Andorra     164           AD               40.0       468   
4       Angola      26           AO               47.5   1246700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2_emissions     cpi  cpi_change currency_code  fertility_rate  \
0         8,672   149.9         2.3           AFN            4.47   
1         4,536  119.05         1.4           ALL            1.62   
2       150,006  151.36         2.0           DZD            3.02   
3           469     NaN         NaN           EUR            1.27   
4        34,693  261.73        17.1           AOA            5.52   

   forested_area  gasoline_price               gdp  \
0            2.1            0.70   19,101,353,833    
1           28.1            1.36   15,278,077,447    
2            0.8            0.28  169,988,236,398    
3           34.0            1.51    3,154,057,987    
4           46.3            0.97   94,635,415,870    

   gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
0                               104.0                                  9.7   
1                               107.0                                 55.0   
2                               109.9                                 51.4   
3                               106.4                                  NaN   
4                               113.5                                  9.3   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio  minimum_wage official_language  \
0                     638.0          0.43            Pashto   
1                      15.0          1.12          Albanian   
2                     112.0          0.95            Arabic   
3                       NaN          6.63           Catalan   
4                     241.0          0.71        Portuguese   

   out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                              78.4                     0.28  38,041,754   
1                              56.9                     1.20   2,854,191   
2                              28.1                     1.72  43,053,054   
3                              36.4                     3.33      77,142   
4                              33.4                     0.21  31,825,295   

   population_labor_force_participation  tax_revenue  total_tax_rate  \
0                                  48.9          9.3            71.4   
1                                  55.7         18.6            36.6   
2                                  41.2         37.2            66.1   
3                                   NaN          NaN             NaN   
4                                  77.5          9.2            49.1   

   unemployment_rate urban_population    latitud   longitud  
0              11.12        9,797,273  33.939110  67.709953  
1              12.33        1,747,593  41.153332  20.168331  
2              11.70       31,510,100  28.033886   1.659626  
3                NaN           67,873  42.506285   1.521801  
4               6.89       21,061,025 -11.202692  17

In [200]:
df_merge = df_merge.astype({'land_area': float})
df_merge.head()

country density abbreviation  agricultural_land  land_area  \
0  Afghanistan      60           AF               58.1   652230.0   
1      Albania     105           AL               43.1    28748.0   
2      Algeria      18           DZ               17.4  2381741.0   
3      Andorra     164           AD               40.0      468.0   
4       Angola      26           AO               47.5  1246700.0   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2_emissions     cpi  cpi_change currency_code  fertility_rate  \
0         8,672   149.9         2.3           AFN            4.47   
1         4,536  119.05         1.4           ALL            1.62   
2       150,006  151.36         2.0           DZD            3.02   
3           469     NaN         NaN           EUR            1.27   
4        34,693  261.73        17.1           AOA            5.52   

   forested_area  gasoline_price               gdp  \
0            2.1            0.70   19,101,353,833    
1           28.1            1.36   15,278,077,447    
2            0.8            0.28  169,988,236,398    
3           34.0            1.51    3,154,057,987    
4           46.3            0.97   94,635,415,870    

   gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
0                               104.0                                  9.7   
1                               107.0                                 55.0   
2                               109.9                                 51.4   
3                               106.4                                  NaN   
4                               113.5                                  9.3   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio  minimum_wage official_language  \
0                     638.0          0.43            Pashto   
1                      15.0          1.12          Albanian   
2                     112.0          0.95            Arabic   
3                       NaN          6.63           Catalan   
4                     241.0          0.71        Portuguese   

   out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                              78.4                     0.28  38,041,754   
1                              56.9                     1.20   2,854,191   
2                              28.1                     1.72  43,053,054   
3                              36.4                     3.33      77,142   
4                              33.4                     0.21  31,825,295   

   population_labor_force_participation  tax_revenue  total_tax_rate  \
0                                  48.9          9.3            71.4   
1                                  55.7         18.6            36.6   
2                                  41.2         37.2            66.1   
3                                   NaN          NaN             NaN   
4                                  77.5          9.2            49.1   

   unemployment_rate urban_population    latitud   longitud  
0              11.12        9,797,273  33.939110  67.709953  
1              12.33        1,747,593  41.153332  20.168331  
2              11.70       31,510,100  28.033886   1.659626  
3                NaN           67,873  42.506285   1.521801  
4               6.89       21,061,025 -11.2026

In [201]:
df_merge = df_merge.astype({'density': float})
df_merge.head()

country  density abbreviation  agricultural_land  land_area  \
0  Afghanistan     60.0           AF               58.1   652230.0   
1      Albania    105.0           AL               43.1    28748.0   
2      Algeria     18.0           DZ               17.4  2381741.0   
3      Andorra    164.0           AD               40.0      468.0   
4       Angola     26.0           AO               47.5  1246700.0   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2_emissions     cpi  cpi_change currency_code  fertility_rate  \
0         8,672   149.9         2.3           AFN            4.47   
1         4,536  119.05         1.4           ALL            1.62   
2       150,006  151.36         2.0           DZD            3.02   
3           469     NaN         NaN           EUR            1.27   
4        34,693  261.73        17.1           AOA            5.52   

   forested_area  gasoline_price               gdp  \
0            2.1            0.70   19,101,353,833    
1           28.1            1.36   15,278,077,447    
2            0.8            0.28  169,988,236,398    
3           34.0            1.51    3,154,057,987    
4           46.3            0.97   94,635,415,870    

   gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
0                               104.0                                  9.7   
1                               107.0                                 55.0   
2                               109.9                                 51.4   
3                               106.4                                  NaN   
4                               113.5                                  9.3   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio  minimum_wage official_language  \
0                     638.0          0.43            Pashto   
1                      15.0          1.12          Albanian   
2                     112.0          0.95            Arabic   
3                       NaN          6.63           Catalan   
4                     241.0          0.71        Portuguese   

   out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                              78.4                     0.28  38,041,754   
1                              56.9                     1.20   2,854,191   
2                              28.1                     1.72  43,053,054   
3                              36.4                     3.33      77,142   
4                              33.4                     0.21  31,825,295   

   population_labor_force_participation  tax_revenue  total_tax_rate  \
0                                  48.9          9.3            71.4   
1                                  55.7         18.6            36.6   
2                                  41.2         37.2            66.1   
3                                   NaN          NaN             NaN   
4                                  77.5          9.2            49.1   

   unemployment_rate urban_population    latitud   longitud  
0              11.12        9,797,273  33.939110  67.709953  
1              12.33        1,747,593  41.153332  20.168331  
2              11.70       31,510,100  28.033886   1.659626  
3                NaN           67,873  42.506285   1.521801  
4               6.89       21,061,025 -1

Ahora hacemos una regla de tres, si tenemos ''x''  muertes por cada 1,000 por km2 nacidos vivos antes de cumplir un año de edad, tendremos que ver cuántas muertes tenemos por el total de habitantes.

¿cuál es el problema? que tenemos habitantes por km2, entonces realmente no nos sirve para nada el lad_area porque ya tenemos la density calculada acorde a los km2. por tanto tendremos que hacer sólo la regla de tres:


|`'num_columna'`  muertes por cada -----> |1,000 nacidos vivos por km2|
|--------------------------|---------------------|
|`x` muertes por cada ---------------------> |total de habitantes por km2|

In [202]:
df_merge['infant_mortality_per_km2'] = round((df_merge["density"] * df_merge['infant_mortality']) / 1000, 2)

In [210]:
df_merge['infant_mortality_per_km2'].describe() #hay algún dato con mortalidad alta pero no sé si saldrá alguno entre 40 y 50, vamos a ver

count    189.000000
mean       3.279524
std        6.749657
min        0.000000
25%        0.390000
50%        1.100000
75%        3.160000
max       68.480000
Name: infant_mortality_per_km2, dtype: float64

In [211]:
df_merge.describe().T.round(2) #si utilizo el .round(2) al describe, me salen los datos redondeados, sin el formato científico: "e+01"

count       mean         std     min  \
density                               195.0     356.76     1982.89    2.00   
agricultural_land                     188.0      39.12       21.78    0.60   
land_area                             194.0  689624.37  1921609.23    0.00   
birth_rate                            189.0      20.21        9.95    5.90   
calling_code                          194.0     360.55      323.24    1.00   
cpi_change                            179.0       6.72       24.45   -4.30   
fertility_rate                        188.0       2.70        1.28    0.98   
forested_area                         188.0      32.02       23.79    0.00   
gasoline_price                        175.0       1.00        0.37    0.00   
gross_primary_education_enrollment    188.0     102.47       13.15   23.40   
gross_tertiary_education_enrollment   183.0      37.96       29.26    0.80   
infant_mortality                      189.0      21.33       19.55    1.40   
life_expectancy                       187.0      72.28        7.48   52.80   
maternal_mortality_ratio              181.0     160.39      233.50    2.00   
minimum_wage                          150.0       2.19        2.97    0.01   
out_of_pocket_health_expenditure      188.0      32.64       19.17    0.20   
physicians_per_thousand               188.0       1.84        1.68    0.01   
population_labor_force_participation  176.0      62.74       10.50   38.00   
tax_revenue                           169.0      16.57        6.96    0.00   
total_tax_rate                        183.0      40.82       20.56    8.00   
unemployment_rate                     176.0       6.89        5.08    0.09   
latitud                               194.0      19.09       23.96  -40.90   
longitud                              194.0      20.23       66.72 -175.20   
infant_mortality_per_km2              189.0       3.28        6.75    0.00   

                                           25%        50%        75%  \
density                                  35.50      89.00     216.50   
agricultural_land                        21.70      39.60      55.38   
land_area                             23828.25  119511.00  524256.00   
birth_rate                               11.30      17.95      28.75   
calling_code                             82.50     255.50     506.75   
cpi_change                                1.00       2.30       4.25   
fertility_rate                            1.70       2.24       3.60   
forested_area                            11.00      32.00      48.18   
gasoline_price                            0.76       0.98       1.24   
gross_primary_education_enrollment       98.95     102.55     108.02   
gross_tertiary_education_enrollment      12.15      31.20      63.15   
infant_mortality                          6.00      14.00      32.70   
life_expectancy                          67.00      73.20      77.50   
maternal_mortality_ratio                 13.00      53.00     186.00   
minimum_wage                              0.40       1.04       2.45   
out_of_pocket_health_expenditure         17.45      30.70      43.98   
physicians_per_thousand                   0.33       1.46       2.94   
population_labor_force_participation     56.15      62.45      69.50   
tax_revenue                              11.70      16.30      21.10   
total_tax_rate                           30.60      37.20      48.00   
unemployment_rate                         3.40       5.36       9.49   
latitud                                   4.54      17.27      40.12   
longitud                                 -7.94      20.97      48.28   
infant_mortality_per_km2                  0.39       1.10       3.16   

                                              max  
density                                  26337.00  
agricultural_land                           82.60  
land_area                             17098240.00  
birth_rate                                  46.08  
calling_code              

In [213]:
#Encuentra todos los países cuya mortalidad infantil esté entre 40 y 50 personas por kilómetro cuadrado.
df_merge[(df_merge['infant_mortality_per_km2'] >= 30) & (df_merge['infant_mortality_per_km2'] <= 70)]

country  density abbreviation  agricultural_land  land_area  \
13   Bangladesh   1265.0           BD               70.6   148460.0   
113      Monaco  26337.0           MC                NaN        2.0   

    armed_forces_size  birth_rate  calling_code capital/major_city  \
13            221,000       18.18         880.0              Dhaka   
113               NaN        5.90         377.0        Monaco City   

    co2_emissions     cpi  cpi_change currency_code  fertility_rate  \
13         84,246  179.68         5.6           BDT            2.04   
113           NaN     NaN         NaN           EUR             NaN   

     forested_area  gasoline_price               gdp  \
13            11.0            1.12  302,571,254,131    
113            NaN            2.00    7,184,844,193    

     gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
13                                116.5                                 20.6   
113                                 NaN                                  NaN   

     infant_mortality largest_city  life_expectancy  maternal_mortality_ratio  \
13               25.1        Dhaka             72.3                     173.0   
113               2.6  Monaco City              NaN                       NaN   

     minimum_wage official_language  out_of_pocket_health_expenditure  \
13           0.51           Bengali                              71.8   
113         11.72            French                               6.1   

     physicians_per_thousand   population  \
13                      0.58  167,310,838   
113                     6.56       38,964   

     population_labor_force_participation  tax_revenue  total_tax_rate  \
13                                   59.0          8.8            33.4   
113                                   NaN          NaN             NaN   

     unemployment_rate urban_population    latitud   longitud  \
13                4.19       60,987,417  23.684994  90.356331   
113                NaN           38,964  43.738418   7.424616   

     infant_mortality_per_km2  
13                      31.75  
113                     68.48

Encuentra los países cuyas tasas de natalidad son mayores o iguales a 20 y su esperanza de vida es mayor de 75 años.

Datos que tenemos:
- Fertility Rate: Número promedio de hijos nacidos de una mujer durante su vida.

- Birth Rate: Número de nacimientos por cada 1,000 habitantes por año.

- Density\n(P/Km2): Densidad de población medida en personas por kilómetro cuadrado.

- Land Area(Km2): Área total de tierra del país en kilómetros cuadrados.

- Life expectancy: Número promedio de años que se espera que viva un recién nacido.


In [220]:
#entiendo que como es nacimientos x cada 1000 hab, tendremos que adaptarlo a la población de cada sitio:

df_merge['birth_rate_total_population'] = (df_merge['life_expectancy'] * df_merge['density']) / 1000

In [221]:
df_merge['birth_rate_total_population'].describe()

count    187.000000
mean      15.958214
std       54.155904
min        0.139400
25%        2.316400
50%        6.232200
75%       14.369250
max      694.549800
Name: birth_rate_total_population, dtype: float64

In [215]:
df_merge['life_expectancy'].describe()

count    187.000000
mean      72.279679
std        7.483661
min       52.800000
25%       67.000000
50%       73.200000
75%       77.500000
max       85.400000
Name: life_expectancy, dtype: float64

In [222]:
#Encuentra los países cuyas tasas de natalidad son mayores o iguales a 20 y su esperanza de vida es mayor de 75 años.

df_merge[(df_merge['birth_rate_total_population'] >= 20) & (df_merge['life_expectancy'] >= 75)]

country  density abbreviation  agricultural_land  land_area  \
12          Bahrain   2239.0           BH               11.1      765.0   
14         Barbados    668.0           BB               23.3      430.0   
16          Belgium    383.0           BE               44.6    30528.0   
82           Israel    400.0           IL               24.6    20770.0   
85            Japan    347.0           JP               12.3   377944.0   
94          Lebanon    667.0           LB               64.3    10400.0   
104        Maldives   1802.0           MV               26.3      298.0   
106           Malta   1380.0           MT               32.4      316.0   
122     Netherlands    508.0           NL               53.3    41543.0   
146     Saint Lucia    301.0           LC               17.4      616.0   
149      San Marino    566.0           SM               16.7       61.0   
156       Singapore   8358.0           SG                0.9      716.0   
162     South Korea    527.0           KR               17.4    99720.0   
165       Sri Lanka    341.0           LK               43.7    65610.0   
185  United Kingdom    281.0           GB               71.7   243610.0   
191         Vietnam    314.0           VN               39.3   331210.0   

    armed_forces_size  birth_rate  calling_code  capital/major_city  \
12             19,000       13.99         973.0              Manama   
14              1,000       10.65           1.0          Bridgetown   
16             32,000       10.30          32.0    City of Brussels   
82            178,000       20.80         972.0           Jerusalem   
85            261,000        7.40          81.0               Tokyo   
94             80,000       17.55         961.0              Beirut   
104             5,000       14.20         960.0                Mal�   
106             2,000        9.20         356.0            Valletta   
122            41,000        9.70          31.0           Amsterdam   
146               NaN       12.00           1.0            Castries   
149               NaN        6.80         378.0  City of San Marino   
156            81,000        8.80          65.0                 NaN   
162           634,000        6.40          82.0               Seoul   
165           317,000       15.83          94.0             Colombo   
185           148,000       11.00          44.0              London   
191           522,000       16.75          84.0               Hanoi   

    co2_emissions     cpi  cpi_change currency_code  fertility_rate  \
12         31,694  117.59         2.1           BHD            1.99   
14          1,276  134.09         4.1           BBD            1.62   
16         96,889  117.11         1.4           EUR            1.62   
82         65,166  108.15         0.8           ILS            3.09   
85      1,135,886  105.48         0.5           NaN            1.42   
94         24,796  130.02         3.0           LBP            2.09   
104         1,445    99.7         0.2           NaN            1.87   
106         1,342  113.45         1.6           EUR            1.23   
122       170,780  115.91         2.6           NaN            1.59   
146           414  110.13         1.9           XCD            1.44   
149           NaN  110.63         1.0           EUR            1.26   
156        37,535  114.41         0.6           SGD            1.14   
162       620,302  115.16         0.4           KRW            0.98   
165        23,362  155.53         3.5           LKR            2.20   
185       379,025  119.62         1.7           GBP            1.68   
191       192,668  163.52         2.8           VND            2.05   

     forested_area  gasoline_price                 gdp  \
12             0.8            0.43     38,574,069,149    
14            14.7            1.81      5,209,000,000    
16            22.6            1.43    529,606,710,418    
82             7.7            1.57    395,098,666,122    
85            68.5            

Encuentra las ciudades cuyos paises contienen la palabra "la" en su nombre.

In [228]:
df_merge.loc[df_merge['country'].str.contains('la', case = False), 'capital/major_city']

4              Luanda
13              Dhaka
15              Minsk
59           Helsinki
68     Guatemala City
76          Reykjav��
81             Dublin
92          Vientiane
93               Riga
102          Lilongwe
103      Kuala Lumpur
107            Majuro
122         Amsterdam
123        Wellington
132         Ngerulmud
139            Warsaw
159           Honiara
165           Colombo
169              Bern
173           Bangkok
190           Caracas
Name: capital/major_city, dtype: object

Encuentra los países cuyos medicos por cada 1000 habitantes (physicians_per_thousand) sea mayores de 5.

- Physicians per thousand: Número de médicos por cada mil personas.

In [230]:
df_merge['physicians_per_thousand'].describe()

count    188.000000
mean       1.839840
std        1.684261
min        0.010000
25%        0.332500
50%        1.460000
75%        2.935000
max        8.420000
Name: physicians_per_thousand, dtype: float64

In [231]:
df_merge.loc[df_merge['physicians_per_thousand'] > 5, 'country']

9         Austria
15        Belarus
42           Cuba
63        Georgia
66         Greece
99      Lithuania
113        Monaco
140      Portugal
149    San Marino
187       Uruguay
Name: country, dtype: object

Encuentra los países cuya tasa de fertilidad sea mayor a 6.

- Fertility Rate: Número promedio de hijos nacidos de una mujer durante su vida.

In [233]:
df_merge.loc[df_merge['fertility_rate'] > 6, 'country']

125      Niger
160    Somalia
Name: country, dtype: object

Encuentra los países cuya moneda es el euro (EUR) y tienen una tasa de natalidad superior al promedio.

- birth_rate_total_population

- currency_code


In [237]:
df_merge.loc[(df_merge['birth_rate_total_population'] >  df_merge['birth_rate_total_population'].mean()), 'country']

5                   Antigua and Barbuda
12                              Bahrain
13                           Bangladesh
14                             Barbados
16                              Belgium
27                              Burundi
38                              Comoros
49                   Dominican Republic
52                          El Salvador
64                              Germany
67                              Grenada
72                                Haiti
77                                India
82                               Israel
83                                Italy
84                              Jamaica
85                                Japan
90                               Kuwait
94                              Lebanon
98                        Liechtenstein
100                          Luxembourg
104                            Maldives
106                               Malta
107                    Marshall Islands
109                           Mauritius


Encuentra los países cuyas tasas de mortalidad infantil son superiores a 70.

In [238]:
df_merge['infant_mortality'].describe()

count    189.000000
mean      21.332804
std       19.548058
min        1.400000
25%        6.000000
50%       14.000000
75%       32.700000
max       84.500000
Name: infant_mortality, dtype: float64

In [239]:
df_merge[df_merge['infant_mortality'] > 70]

country  density abbreviation  agricultural_land  \
33   Central African Republic      8.0           CF                8.2   
34                       Chad     13.0           TD               39.7   
126                   Nigeria    226.0           NG               77.7   
155              Sierra Leone    111.0           SL               54.7   
160                   Somalia     25.0           SO               70.3   

     land_area armed_forces_size  birth_rate  calling_code capital/major_city  \
33    622984.0             8,000       35.35         236.0             Bangui   
34   1284000.0            35,000       42.17         235.0          N'Djamena   
126   923768.0           215,000       37.91         234.0              Abuja   
155    71740.0             9,000       33.41         232.0           Freetown   
160   637657.0            20,000       41.75         252.0          Mogadishu   

    co2_emissions     cpi  cpi_change currency_code  fertility_rate  \
33            297  186.86        37.1           NaN            4.72   
34          1,016   117.7        -1.0           XAF            5.75   
126       120,369  267.51        11.4           NGN            5.39   
155         1,093  234.16        14.8           SLL            4.26   
160           645     NaN         NaN           SOS            6.07   

     forested_area  gasoline_price               gdp  \
33            35.6            1.41    2,220,307,369    
34             3.8            0.78   11,314,951,343    
126            7.2            0.46  448,120,428,859    
155           43.1            1.08    3,941,474,311    
160           10.0            1.41    4,720,727,278    

     gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
33                                102.0                                  3.0   
34                                 86.8                                  3.3   
126                                84.7                                 10.2   
155                               112.8                                  2.0   
160                                23.4                                  2.5   

     infant_mortality largest_city  life_expectancy  maternal_mortality_ratio  \
33               84.5       Bangui             52.8                     829.0   
34               71.4    N'Djamena             54.0                    1140.0   
126              75.7        Lagos             54.3                     917.0   
155              78.5     Freetown             54.3                    1120.0   
160              76.6       Bosaso             57.1                     829.0   

     minimum_wage official_language  out_of_pocket_health_expenditure  \
33           0.37            French                              39.6   
34           0.60            French                              56.4   
126          0.54           English                              72.2   
155          0.57           English                              38.2   
160           NaN            Arabic                               NaN   

     physicians_per_thousand   population  \
33                      0.06    4,745,185   
34                      0.04   15,946,876   
126                     0.38  200,963,599   
155                     0.03    7,813,215   
160                     0.02   15,442,905   

     population_labor_force_participation  tax_revenue  total_tax_rate  \
33                                   72.0          8.6            73.3   
34                                   70.7          NaN            63.5   
126                                  52.9          1.5            34.8   
155                                  57.9          8.6            30.7   
160                                  47.4          0.0             NaN   

     unemployment_rate urban_population    latitud   longitud  \
33                3.68        1,982,064   6.611111  20.939444   
34                1.89        3,712,273  15.454166  18.732207   
126     